File: akka-basic/src/main/java/org/darion/yaphet/java/akka/hotswap/HotSwapActor.java
package org.darion.yaphet.java.akka.hotswap;

import akka.actor.AbstractActor;

public class HotSwapActor extends AbstractActor {
    private AbstractActor.Receive angry;
    private AbstractActor.Receive happy;

    public HotSwapActor() {
        angry = receiveBuilder()
                .matchEquals("foo", s -> getSender().tell("I am already angry?", getSelf()))
                .matchEquals("bar", s -> getContext().become(happy))
                .build();

        happy = receiveBuilder()
                .matchEquals("bar", s -> getSender().tell("I am already happy :-)", getSelf()))
                .matchEquals("foo", s -> getContext().become(angry))
                .build();
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .matchEquals("foo", s -> getContext().become(angry))
                .matchEquals("bar", s -> getContext().become(happy))
                .build();
    }
}

File: akka-basic/src/main/scala/org/darion/yaphet/akka/hotswap/HotSwapActor.scala
package org.darion.yaphet.akka.hotswap

import akka.actor.Actor

class HotSwapActor extends Actor {

  import context._

  def angry: Receive = {
    case "foo" ⇒ sender() ! "I am already angry?"
    case "bar" ⇒ become(happy)
  }

  def happy: Receive = {
    case "bar" ⇒ sender() ! "I am already happy :-)"
    case "foo" ⇒ become(angry)
  }

  override def receive: Receive = {
    case "foo" ⇒ become(angry)
    case "bar" ⇒ become(happy)
  }
}
